AzureCreateVirtualMachineTask: introduce new task

nirarg opened the pull request:
Add new task creating VM on Azure cloud
Currently it create Ubuntu

<!--
- Please ensure code changes are split into a series of logically independent commits.
- Every commit should have a subject/title (What) and a description/body (Why).
- Every PR must have a description.
- As an example you can use git commit -m"What" -m"Why" to achieve the requirements above. GitHub automatically recognises the commit description (-m"Why") in single commit PRs and adds it as the PR description.
- If the PR includes auto generated code, make sure to add this code in different commit (last commit), to make the review process easier.
-->

**What this PR does / why we need it**:

This PR is created as part of the next demo requirements

**Which issue(s) this PR fixes*:
Fixes #[FLPATH-362](https://issues.redhat.com/browse/FLPATH-362)

**Change type**
- [x] New feature
- [ ] Bug fix
- [ ] Unit tests
- [ ] Integration tests
- [ ] CI
- [ ] Documentation
- [ ] Auto-generated SDK code

**Impacted services**
- [x] Workflow Service
- [ ] Notification Service

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.

masayag commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,147 @@
+import org.springframework.util.StringUtils;
+
+@Slf4j
+public class AzureCreateVirtualMachineTask extends BaseInfrastructureWorkFlowTask {
```
thinking if `Azure` prefix is needed for the class name if it is already placed in `.azure` package?

nirarg replied:
This is the naming convention used by all tasks under Prebuild module
Do you think we should change it?

masayag replied:
don't have a strong opinion on either way.

masayag commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,147 @@
+import com.azure.core.management.AzureEnvironment;
+import com.azure.core.management.Region;
+import com.azure.core.management.profile.AzureProfile;
+import com.azure.identity.*;
```
pls expand.

masayag commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,147 @@
+
+			String ipAddress = publicIpAddress.ipAddress();
+			log.info("VirtualMachine was created with public IP {}", ipAddress);
+			context.put("public-ip-address", ipAddress);
```
Can this be described somewhere? Seems like a part of the task's output.

nirarg replied:
Added description to the execute function, any other idea how to describe this?

masayag commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,147 @@
+					.create();
+
+			PublicIpAddress publicIpAddress = virtualMachine.getPrimaryPublicIPAddress();
+			if (publicIpAddress == null) {
```
Is it a simple getter? can the IP address be configured after the VM is actually up and running and obtained the address?

nirarg replied:
The public IP is created before the Virtual machine created, therefore it shouldn't be empty at this point.
I added this validation just to be on the safe side

RichardW98 commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,147 @@
+			VirtualMachine virtualMachine = azureResourceManager.virtualMachines().define(resourcesPrefix + "VM")
+					.withRegion(Region.US_EAST).withExistingResourceGroup(resourcesPrefix + "ResourceGroup")
+					.withExistingPrimaryNetworkInterface(networkInterface)
+					.withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_18_04_LTS)
```
should we make this dynamic to support windows server or other systems?

masayag replied:
+1, I'd suggest to parameterized it (or a closed list of options) + also let the user specify VirtualMachineSizeTypes

nirarg replied:
I suggest to have this change after the demo

nirarg replied:
As from the first impression I got, its not just about changing the machine type, also the function (withPopularLinuxImage) and the size might be changed .
Need to investigate more about the types included.

pkliczewski commented on prebuilt-tasks/pom.xml:
```diff
@@ -134,5 +134,22 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-core</artifactId>
+            <version>1.37.0</version>
```
please use properties for versions like we do it for other dependencies.

nirarg replied:
Done

pkliczewski commented on prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:
```diff
@@ -0,0 +1,153 @@
+		catch (MissingParameterException e) {
+			log.error("Task {} failed: missing required parameter, error: {}", getName(), e.getMessage());
+		}
+		catch (Exception e) {
```
please be specific which exceptions we want to catch here

nirarg replied:
Except to `MissingParameterException` there is no known declared Exception that should be catched
Therefore I catch `Exception` to cover any unexpected `RuntimeException`

pkliczewski reviewed (commented):
please add some tests

nirarg commented:
> please add some tests

@pkliczewski 
I'm working on adding unit tests for the new task
Maybe would add this in new PR, in order to have this PR merged before the demo

pkliczewski reviewed (commented):
/lgtm

nirarg commented:
Added unit tests to the PR
Added all new changes in different commit: https://github.com/parodos-dev/parodos/pull/346/commits/887c63308521e5ebc45cdca87babd2228962fff2

masayag commented on prebuilt-tasks/src/test/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTaskTest.java:
```diff
@@ -0,0 +1,140 @@
+
+	@Test
+	public void testHappyFlow() {
+		String publicIP = "11.11.11.11";
```
pls add tests using the convention:
```
/ /given
...
// when
..
// then
..
```
see as en example https://github.com/parodos-dev/parodos/blob/main/workflow-service/src/test/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImplTest.java#L175

nirarg replied:
Done

nirarg commented:
Also added instructions how to use the example workflow configuration
see [here](https://github.com/parodos-dev/parodos/pull/346/commits/463e04a2de9f25f7855f4977d0224ff5b400f9db)

masayag commented:
/lgtm
/approve
